.circleci/config.yml

```diff
@@ -1,15 +1,131 @@
+    attach_workspace:
+      at: *workspace
+
+  save_workspace: &save_workspace
+    persist_to_workspace:
+      root: *workspace
+      paths:
+      - farmsmart
+
 jobs:
-  build:
+
+  decrypt-secrets:
     docker:
-      - image: cirrusci/flutter
+      - image: farmsmart/build-agent
+    working_directory: *working_dir
     steps:
       - checkout
+      - run: |
+          SECRETS_DIR=./creds
+          mkdir -p ${SECRETS_DIR}
+          echo ${GOOGLE_SERVICE_ACCOUNT} | base64 -di | tee ${SECRETS_DIR}/service-account.json | gcloud auth activate-service-account --key-file=-
+          export GOOGLE_APPLICATION_CREDENTIALS=${SECRETS_DIR}/service-account.json
+          ./secrets/sops-decrypt.sh
+      - *save_workspace
+
+  get-next-version:
+    docker:
+      - image: circleci/python:3.7.3
+    working_directory: *working_dir
+    steps:
+      - *restore_workspace
+      - run: |
+          if echo $CIRCLE_BRANCH | grep -E -i -q "release"
+            then
+              BRANCH=release
+            else
+              BRANCH=develop
+          fi
+          source ./ci/env.${BRANCH}.sh
+          sudo pip3 install -r ./ci/playstore/requirements.txt
+          VERSION_INFO=$(sudo python3 ./ci/playstore/get_next_version.py co.farmsmart.app ${VERSION} ${TRACK})
+          VERSION_NAME=$(echo ${VERSION_INFO} | jq '.version_name')
+          VERSION_CODE=$(echo ${VERSION_INFO} | jq '.version_code')
+          sudo echo "export VERSION_NAME=${VERSION_NAME}" >> ./env.sh
+          sudo echo "export VERSION_CODE=${VERSION_CODE}" >> ./env.sh
+          sudo cat ./ci/env.${BRANCH}.sh >> ./env.sh
+      - *save_workspace
+
+  flutter-build:
+    docker:
+      - image: cirrusci/flutter:v1.2.1
+    working_directory: *working_dir
+    steps:
+      - *restore_workspace
+      - run: |
+          source ./env.sh
+          echo "Building flutter apk - BUILD_NAME:${VERSION_NAME}, BUILD_NUMBER=${VERSION_CODE}"
+          flutter build apk --build-name=${VERSION_NAME} --build-number=${VERSION_CODE} --release --flavor production
+      - *save_workspace
+
+  promote-apk-to-track:
+      docker:
+        - image: circleci/python:3.7.3
+      working_directory: *working_dir
+      steps:
+        - *restore_workspace
+        - run: |
+            source ./env.sh
+            sudo pip3 install -r ./ci/playstore/requirements.txt
+            sudo python3 ./ci/playstore/promote_apk_to_track.py co.farmsmart.app ${TRACK} ${VERSION_NAME} ${VERSION_CODE} ./Release_notes build/app/outputs/apk/production/release/app-production-release.apk
+        - *save_workspace
+
+  slack-notification:
+      docker:
+        - image: circleci/python:3.7.3
+      working_directory: *working_dir
+      steps:
+        - *restore_workspace
+        - run: |
+            source ./env.sh
+            sudo pip3 install -r ./ci/playstore/requirements.txt
+            sudo python3 ./ci/playstore/notify.py ${SLACK_FARMSMART_TOKEN} ${TRACK} ${TRACK_URL} ${VERSION_NAME}
```

Comment:
Can we define a circle ci **context** for this so that some environment variables is possible for reuse such as slack and not just this repo? 
I'd be good to use the same parameters for when we have our firebase-backend slack comms.

One advantage of context is that it has security rules of who are allowed to use them in scripts

Reply:
We can use contexts for global variables - I'll update this in CircleCI.

However you aren't able to define context variables per a branch. It's possible to have two different jobs which run with different contexts however it will lead to duplication:

All steps would need to duplicated per environment which isn't the most readable:

jobs:
      - flutter-build-dev:
          run: |
            ....
          context: dev

      - flutter-build-prod:
          run: |
            ....
          context: prod

workflows:
  run:
    jobs:
      - flutter-build-dev:
          filters:
            branches:
              only: /develop/
      - flutter-build-prod:
          filters:
            branches:
              only: /release.*/


This is why I went down the route of dynamically loading environment variables from the repo.

We can certainly abstract any global variables into a context though to share across repos.

Reply:
can you apply the context in the workflow instead? that way you just adjust the workflow?

Reply:
It's the approach I did for the firebase-backend. the only bit i redefined a workflow step for deploy.

Reply:
Please see: https://github.com/farmsmart/farmsmart-firebase/blob/develop/.circleci/config.yml

Reply:
I've now updated the workflow to use contexts